### PR TITLE
validate input in register default constructor

### LINKED ISSUE #397: add input validation to register default constructor
`Register.__init__` lacks input validation

**File:** `qoolqit/register.py` — `Register.__init__`

## Problems

1. **No empty dict guard**: `Register({})` is accepted silently, but `min_distance()`, `max_radial_distance()`, and `draw()` all crash downstream with unhelpful errors.

2. **No coordinate value validation**: values are never checked to be 2-tuples of numbers. Inputs like `{"a": "bad"}` or `{"a": (1, 2, 3)}` pass construction and cause cryptic errors later.

3. **Missing generic type annotation**: `self._qubits: dict` gives no type information to static checkers. Should be `dict[Any, tuple[float, float]]` or a named alias.

### PR DESCRIPTION
**Pull request by sgrava**
Resolves #397 

> `Register.__init__` lacks input validation
> 
> **File:** `qoolqit/register.py` — `Register.__init__`
> 
> ## Problems
> 
> 1. **No empty dict guard**: `Register({})` is accepted silently, but `min_distance()`, `max_radial_distance()`, and `draw()` all crash downstream with unhelpful errors.
> 
> 2. **No coordinate value validation**: values are never checked to be 2-tuples of numbers. Inputs like `{"a": "bad"}` or `{"a": (1, 2, 3)}` pass construction and cause cryptic errors later.
> 
> 3. **Missing generic type annotation**: `self._qubits: dict` gives no type information to static checkers. Should be `dict[Any, tuple[float, float]]` or a named alias.
> 

## Changes
- expand register docstrings
- validate input, we accept all cordinates that can be casted to numpy or torch array of shape (2,)

## Next
Next steps are collected in #388 
- validate input and small fixes in Register.from_coordinates()
- deprecate Register.from_graph()

### CODE REVIEW

File: qoolqit/register.py [hunk 1]
```diff
@@ -1,28 +1,93 @@
...
         if not isinstance(qubits, dict):
-            raise TypeError(
-                "Register must be initialized with a dictionary of "
-                "qubits and respective coordinates {q: (x, y), ...}."
-            )
-
-        self._qubits: dict = qubits
+            raise TypeError("`qubits` must be a dictionary mapping qubit ids to coordinates.")
+        if len(qubits) == 0:
+            raise ValueError("Register cannot be empty.")
```
Comment (abussy-pasqal) [inline qoolqit/register.py:100]: I can't see why a register could not be empty. To me, it seems that a register is simply a decorated `dict`. So i would try to implement every function that is already implemented as its `dict` counterpart.

I'll add comments where relevent.

This raises the question of `Register` being only an alias of a `dict`, but this would be too huge a refactoring here

File: qoolqit/register.py [hunk 2]
```diff
@@ -49,15 +114,16 @@ def from_coordinates(cls, coords: list) -> Register:
...
             )
         coords_dict = {i: pos for i, pos in enumerate(coords)}
         return cls(coords_dict)
 
     @property
     def qubits(self) -> dict:
-        """Returns the dictionary of qubits and respective coordinates."""
-        return self._qubits
+        """Returns a copy of the dictionary of qubits and respective coordinates."""
+        return dict(self._qubits)
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: Please add a test that the 2D-vectors are indeed copied. Pseudo-code :

```python
qubits1 = register.qubits
qubits2 = register.qubits
qubits1["0"][0] = 1.0
assert qubits1 != qubits2
```

I suspect that this test will fail.

Avoid using `copy.deepcopy`. Use :
```python
import numpy as np
import torch
from copy import deepcopy

def copy_value(v):
    if isinstance(v, np.ndarray):
        return v.copy()
    if isinstance(v, torch.Tensor):
        # Don't use .detach() if you want to preserve autograd history,
        # which is probably not the case here
        return v.detach().clone()
    return deepcopy(v)

copied = {k: copy_value(v) for k, v in self._qubits.items()}
```

File: qoolqit/register.py [hunk 3]
```diff
@@ -67,7 +133,7 @@ def qubits_ids(self) -> list:
     @property
     def n_qubits(self) -> int:
         """Number of qubits in the Register."""
-        return len(self.qubits)
+        return len(self._qubits)
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: I would implement both `__bool__` and `__len__`:
```python
def __bool__(self) -> bool:
    return bool(self._qubits)

def __len__(self) -> int:
    return len(self._qubits)
```
This property would become
```python
def n_qubits(self) -> int:
    return len(self)
```

File: qoolqit/register.py [hunk 4]
```diff
@@ -1,28 +1,93 @@
...
+        if len(qubits) == 0:
+            raise ValueError("Register cannot be empty.")
+
+        parsed = {}
+        for key, val in qubits.items():
+            try:
+                if _has_torch and isinstance(val, torch.Tensor):
+                    parsed[key] = torch.asarray(val, dtype=torch.float64)
+                else:
+                    parsed[key] = np.asarray(val, dtype=float)
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: why keeping points as `torch` and not converting everything to `numpy` ? This would save a lot of double handling

Comment (abussy-pasqal) [reply qoolqit/register.py]: TBH, from a geometrical point of view, I would rather store a list of labels and a Nx2 `np.array` (vertices).
This would simplify a lot of the code :  most of (if not all) computations would become simple `numpy` functions

File: qoolqit/register.py (hunk 1 above)
Comment (sgrava) [reply qoolqit/register.py:100]: > I can't see why a register could not be empty. To me, it seems that a register is simply a decorated dict. So i would try to implement every function that is already implemented as its dict counterpart.

I gave it a lot of thoughts about this.
Ultimately, my take is that an empty register is an invalid state (for what we want to achieve here), and shouldn't be propagated down the stack. 
I am happy to rethink about it if a concrete use case pops up.  What do you think?

File: qoolqit/register.py (hunk 2 above)
Comment (sgrava) [reply qoolqit/register.py]: Nice suggestion! 
I added a `_copy` method and the the test you suggested.

File: qoolqit/register.py (hunk 4 above)
Comment (sgrava) [reply qoolqit/register.py]: > why keeping points as torch and not converting everything to numpy ? This would save a lot of double handling

We want to keep the option to pass torch tensor as other users might come from there.
It was already allowed by the register constructor, just not validated, which is the point of this PR.
I'd like to keep it for now. Thanks to your suggestions is manageble and can be refactored easily.

> TBH, from a geometrical point of view, I would rather store a list of labels and a Nx2 np.array (vertices).
This would simplify a lot of the code :  most of (if not all) computations would become simple numpy functions

Good point. 
I store cordinates as an array now. 
- We can think later about touching the default constructor. A better solution would be to have `Register.from_dict` and `Register.from_coords`. Would you agree?
- we can think later about better leveraging this within the class methods. (This PR is already huge.)

File: qoolqit/register.py (hunk 3 above)
Comment (sgrava) [reply qoolqit/register.py]: Done, have a look!

File: qoolqit/register.py (hunk 1 above)
Comment (abussy-pasqal) [reply qoolqit/register.py:100]: It’s a perfectly valid choice.

That said, I think the two points you mentioned are worth distinguishing:

1. **“An empty register is an invalid state”**: this could be represented with a `__bool__` operator that returns `False` if and only if the register is empty.

2. **“It shouldn’t be propagated down the stack”**: one could argue that a register is a container, much like a `dict`. For some algorithms, an empty `dict` would be invalid input; for others, it would simply be a no-op. In my opinion, that decision should be left to the consumer.

So I agree that, in the absence of a concrete use case, enforcing the check in the constructor is probably the more convenient and safer option.

File: qoolqit/register.py [hunk 5]
```diff
@@ -1,34 +1,150 @@
...
+        return x.clone()
+    return np.copy(x)
 
 
 class Register:
-    """The Register in QoolQit, representing a set of qubits with coordinates."""
+    """A QoolQit register mapping qubit IDs to 2D coordinates.
+
+    Attributes:
+        qubits: a dictionary of qubit IDs and respective coordinates {q: (x, y),...}.
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: Shouldn't it be removed ?

File: qoolqit/register.py [hunk 6]
```diff
@@ -1,34 +1,150 @@
...
+
+        >>> import torch
+        >>> reg = Register({"a": torch.tensor([0.0, 0.0]), "b": torch.tensor([1.0, 0.0])})
+    """
+
+    def __init__(
+        self,
+        qubits: (
+            Mapping[str, Sequence[float] | npt.NDArray[np.float64] | torch.Tensor]
+            | Mapping[int, Sequence[float] | npt.NDArray[np.float64] | torch.Tensor]
```
Comment (abussy-pasqal) [inline qoolqit/register.py:82]: why the duplication ?

File: qoolqit/register.py [hunk 7]
```diff
@@ -1,34 +1,150 @@
...
-                "Register must be initialized with a dictionary of "
-                "qubits and respective coordinates {q: (x, y), ...}."
-            )
+        if not isinstance(qubits, Mapping):
+            raise TypeError("`qubits` must be a Mapping of qubit ids to coordinates.")
+        if not qubits:
+            raise ValueError("Register cannot be empty.")
+
+        self._qubits_ids: tuple[str | int, ...] = tuple(qubits.keys())
+        validated_coords = [self._validate_coord(k, c) for k, c in qubits.items()]
```
Comment (abussy-pasqal) [inline qoolqit/register.py:103]:
```suggestion
        validated_coords = [self._validate_coord(k, c).tolist() for k, c in qubits.items()]
```

Comment (abussy-pasqal) [reply qoolqit/register.py:103]: This would simplify the code of `_stack_coords`

File: qoolqit/register.py [hunk 8]
```diff
@@ -109,13 +237,12 @@ def draw(self, ax: Axes | None = None, marker_size: int = 100) -> None:
         marker_radius = marker_size**0.5 / 2  # in points
         annotation_offset = 1.5 * marker_radius  # place label just outside the marker
 
-        x, y = zip(*self.qubits.values())
-        ax.scatter(x, y, s=marker_size, color="green")
-
-        for i, qubit_id in enumerate(self.qubits_ids):
+        coords = self._coords.detach().cpu().numpy() if _is_torch(self._coords) else self._coords
+        for xi, yi, qid in zip(coords[:, 0], coords[:, 1], self.qubits_ids):
+            ax.scatter(xi, yi, s=marker_size, color="green")
```
Comment (abussy-pasqal) [inline qoolqit/register.py:239]: Note that `scatter` accepts an array so you could do

```python
coords = self._coords.detach().cpu().numpy() if _is_torch(self._coords) else self._coords

ax.scatter(
    coords[:, 0],
    coords[:, 1],
    s=marker_size,
    color="green",
)
```
But `annotate` doesn't...

File: qoolqit/register.py [hunk 9]
```diff
@@ -1,34 +1,150 @@
...
+def _norm(x: npt.NDArray[np.float64] | torch.Tensor) -> np.float64 | torch.Tensor:
+    if _is_torch(x):
+        norm: torch.Tensor = torch.linalg.norm(x)
+        return norm
+    return np.linalg.norm(x)
+
+
+def _copy(x: npt.NDArray | torch.Tensor) -> npt.NDArray | torch.Tensor:
+    if _is_torch(x):
+        return x.clone()
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: add a a comment why you don't use `detach()` if that's intended

File: qoolqit/register.py (hunk 4 above)
Comment (abussy-pasqal) [reply qoolqit/register.py]: Totally agree 👍

File: qoolqit/register.py (hunk 5 above)
Comment (sgrava) [reply qoolqit/register.py]: You are right, removed them. Just delegate to property docstrings.

File: qoolqit/register.py (hunk 8 above)
Comment (sgrava) [reply qoolqit/register.py:239]: Unfortunately, `annotate` is not vectorized.

File: qoolqit/register.py (hunk 9 above)
Comment (sgrava) [reply qoolqit/register.py]: You are right, added a detach.

File: qoolqit/register.py (hunk 6 above)
Comment (sgrava) [reply qoolqit/register.py:82]: I'm not a typing expert but to me those are just hints to the the user.
It is saying don't do weird combination of labels like [1, "q", 12, ...], just stick to a single type.

File: qoolqit/register.py [hunk 10]
```diff
@@ -1,34 +1,145 @@
...
+    """
+    if _is_torch(val):
+        return val.to(dtype=torch.float64)
+    return np.asarray(val, dtype=np.float64)
+
+
+def _norm(x: npt.NDArray[np.float64] | torch.Tensor) -> np.float64 | torch.Tensor:
+    if _is_torch(x):
+        norm: torch.Tensor = torch.linalg.norm(x)
+        return norm
```
Comment (abussy-pasqal) [inline qoolqit/register.py]:
```suggestion
        return norm.item()
```

File: qoolqit/register.py (hunk 7 above)
Comment (sgrava) [reply qoolqit/register.py:103]: You'll lose the numpy/torch information this way.
I think the best way to simplify here would be to make `_to_array`  bit smarter and recursively call itself to end up with an array idependently of the input.

File: qoolqit/register.py (hunk 6 above)
Comment (abussy-pasqal) [reply qoolqit/register.py:82]: Sorry, I misread the nesting of `[]`. Maybe it's worth doing type aliases ?
```python
from typing import TypeAlias

Array: TypeAlias = npt.NDArray[np.float64] | torch.Tensor
ArrayLike: TypeAlias = npt.ArrayLike | Array
```
Since it's used a lot.

File: qoolqit/register.py (hunk 10 above)
Comment (abussy-pasqal) [reply qoolqit/register.py]: With this, the function always returns a float (maybe you'll need casting)

**Comment by sgrava** [comment]
@abussy-pasqal 

> Maybe it's worth doing type aliases ?
> 
> from typing import TypeAlias
> 
> Array: TypeAlias = npt.NDArray[np.float64] | torch.Tensor
> ArrayLike: TypeAlias = npt.ArrayLike | Array
> 
> Since it's used a lot.

Yes, let's explore it in #406

**Review by abussy-pasqal — DISMISSED** [review]
Feel free to discard the last comment

File: qoolqit/register.py (hunk 10 above)
Comment (sgrava) [reply qoolqit/register.py]: Done! Yes item is typed loosely. I cast to float.

File: qoolqit/register.py [hunk 11]
```diff
@@ -39,11 +39,11 @@ def _to_array(
...
 
 
-def _norm(x: npt.NDArray[np.float64] | torch.Tensor) -> np.float64 | torch.Tensor:
+def _norm(x: npt.NDArray[np.float64] | torch.Tensor) -> float:
     if _is_torch(x):
-        norm: torch.Tensor = torch.linalg.norm(x)
-        return norm
-    return np.linalg.norm(x)
+        norm = torch.linalg.norm(x)
+    norm = np.linalg.norm(x)
```
Comment (abussy-pasqal) [inline qoolqit/register.py]: `np.linalg.norm` is also called in the `torch` branch

**Review by abussy-pasqal — CHANGES_REQUESTED** [review]

**Review by abussy-pasqal — APPROVED** [review]

File: qoolqit/register.py (hunk 11 above)
Comment (sgrava) [reply qoolqit/register.py]: ah my bad, thanks!
